custom/enikshay/reports/views.py

```diff
@@ -57,45 +59,65 @@ class DistrictLocationsView(LocationsView):
-            case['device_number_in_form'] = form_device_number
-
-            user = CommCareUser.get(form.user_id) if form.user_id else None
-            if user:
-                case['username'] = user.username
+    def get(self, request, domain, case_type, *args, **kwargs):
+        self.case_type = {'voucher': CASE_TYPE_VOUCHER, 'person': CASE_TYPE_PERSON}[case_type]
+        self.accessor = CaseAccessors(domain)
+        case_ids = self.accessor.get_case_ids_in_domain(self.case_type)
+        bad_cases = self.get_cases_with_duplicate_ids(case_ids)
+
+        for case in bad_cases:
+            form = CaseAccessorSQL.get_transactions(case['case_id'])[0].form
+            if form:
+                case['form_name'] = form.form_data.get('@name', 'NA')
+                form_device_number = form.form_data.get('serial_id', {}).get('outputs', {}).get('device_number')
+                case['device_number_in_form'] = form_device_number
+                case['form_device_id'] = form.metadata.deviceID
+                case['form_user_id'] = form.user_id
+
+        user_info = self.get_user_info(case['form_user_id'] for case in bad_cases
+                                       if 'form_user_id' in case)
+        for case in bad_cases:
+            user_dict = user_info.get(case.get('form_user_id'))
+            if user_dict:
+                case['username'] = user_dict['username']
                 try:
-                    device_number = [d.device_id for d in user.devices].index(form.metadata.deviceID) + 1
+                    device_number = user_dict['device_ids'].index(case['form_device_id']) + 1
                 except ValueError:
                     device_number = -1
                 case['real_device_number'] = unicode(device_number)
 
-    context = {
-        'case_type': case_type,
-        'num_bad_cases': len(bad_cases),
-        'num_total_cases': len(all_cases),
-        'num_good_cases': len(all_cases) - len(bad_cases),
-        'bad_cases': sorted(bad_cases, key=lambda case: case['opened_on'], reverse=True)
-    }
-    return render(request, 'enikshay/duplicate_ids_report.html', context)
+        context = {
+            'case_type': self.case_type,
+            'num_bad_cases': len(bad_cases),
+            'num_total_cases': len(case_ids),
+            'num_good_cases': len(case_ids) - len(bad_cases),
+            'bad_cases': sorted(bad_cases, key=lambda case: case['opened_on'], reverse=True)
+        }
+        return render(request, 'enikshay/duplicate_ids_report.html', context)
+
+    def get_cases_with_duplicate_ids(self, all_case_ids):
+        id_property = {'voucher': VOUCHER_ID, 'person': ENIKSHAY_ID}[self.case_type]
+        all_cases = [
+            {
+                'case_id': case.case_id,
+                'readable_id': case.get_case_property(id_property),
+                'opened_on': case.opened_on,
+            }
+            for case in self.accessor.iter_cases(all_case_ids)
```

Comment:
wowza. Could you have used a document aggregation / top hits aggregation on the case search index for this instead? It might be a bit harder with the nested documents, but might mean that you don't have to pull every single case into memory.

Reply:
Probably - this was purely a ease-of-implementation decision.  I will note that it doesn't pull them all into memory at once, just in chunks.  The speed it takes to generate this report doesn't matter at all (and the rest of this is also pretty slow), so I just haven't prioritized figuring out how to do that.

Reply:
ok